### PR TITLE
measure supports cols arg

### LINKED ISSUE #5063: Odd behaviour of `melt()` when using the `cols` argument in `measure()`
`measure()` is a great enhancement to `melt()`  - thanks a lot for this!

However, while experimenting with the new `measure()` function when trying to find an answer for [this question on SO](https://stackoverflow.com/q/68233050/3817004) I came across an odd behaviour of `melt()` when the `cols` argument is used with `measure()`:

`#` [`Minimal reproducible example`](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example)

```
df <- fread("Country    AXOX1991    AXOX1992    BXOX1991    BXOX1992    CXOX1991    CXOX1992
Afghanistan 1   2   3   4   5   6
USA 6   5   4   3   2   1")
```

This gives the expected result (note that the `AXO` columns intentionally have be left out of the reshape):

```
melt(df,  measure.vars = measure(var, year, 
                                 pattern = "([BC]XO)X(\\d{4})"))
```
```
       Country AXOX1991 AXOX1992 var year value
1: Afghanistan        1        2 BXO 1991     3
2:         USA        6        5 BXO 1991     4
3: Afghanistan        1        2 BXO 1992     4
4:         USA        6        5 BXO 1992     3
5: Afghanistan        1        2 CXO 1991     5
6:         USA        6        5 CXO 1991     2
7: Afghanistan        1        2 CXO 1992     6
8:         USA        6        5 CXO 1992     1
```

Now, when I include `cols` in the call to `measure()`

```
(cols <- names(df)[4:7])

[1] "BXOX1991" "BXOX1992" "CXOX1991" "CXOX1992"

melt(df,  measure.vars = measure(var, year, cols = cols,
                                 pattern = "([BC]XO)X(\\d{4})"))
```

I get an unexpected and obviously wrong result and a warning:

```
   BXOX1992 CXOX1991 CXOX1992 var year       value
1:        4        5        6 BXO 1991 Afghanistan
2:        3        2        1 BXO 1991         USA
3:        4        5        6 BXO 1992           1
4:        3        2        1 BXO 1992           6
5:        4        5        6 CXO 1991           2
6:        3        2        1 CXO 1991           5
7:        4        5        6 CXO 1992           3
8:        3        2        1 CXO 1992           4
```

> Warning message:
> In melt.data.table(df, measure.vars = measure(var, year,  :
>   'measure.vars' [Country, AXOX1991, AXOX1992, BXOX1991, ...] are not all of the same type. By order of hierarchy, the molten data value column will be of type 'character'. All measure variables not of type 'character' will be coerced too. Check DETAILS in ?melt.data.table for more on coercion.
> 

The documentation says on the `pattern` argument:

> Perl-compatible regex with capture groups to match to `cols`.

Perhaps, I have completely misunderstood the meaning of the `cols` argument. Unfortunately, I have found no example of the proper use of `cols` neither in the documentation of `measure()` nor in the reshape vignette.

--------------

Even stranger, when using a function to create the _character vector of column names_ in the call, I get an error message right away:

```
melt(df,  measure.vars = measure(var, year, cols = names(df)[4:7],
                                 pattern = "([BC]XO)X(\\d{4})"))
``` 

> Error in (function (fun.list, sep = "_", pattern, cols, multiple.keyword = "value.name",  : 
>   cols must be a character vector of column names

`#` `Output of sessionInfo()`

```
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.1

loaded via a namespace (and not attached):
[1] compiler_4.1.0 tools_4.1.0  
```

### PR DESCRIPTION
**Pull request by tdhock**
Closes #5063 in which some users wanted to provide the cols arg to measure, which may be useful/convenient instead of using pattern arg.

This is an alternative to #5115 which would make using the cols arg an error. Moving forward with that other PR would be complicated since there is quite a bit of code which is using the cols arg with patterns, thanks @MichaelChirico for investigating.

So this PR may be considered if we want to add support for cols arg (seems like users want it in #5063 at least).

### CODE REVIEW

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -3159,6 +3159,9 @@ test(1034, as.data.table(x<-as.character(sample(letters, 5))), data.table(V1=x))
   DT_missing_l_2 = data.table(num_1=1, num_2=2, list_1=list(1), list_3=list(3))
   test(1035.0186, melt(DT_missing_l_2, measure.vars=measure(value.name, char, sep="_"), na.rm=TRUE), data.table(char="1", num=1, list=list(1)))
   test(1035.0187, melt(DT_missing_l_2, measure.vars=measure(value.name, char, sep="_"), na.rm=FALSE), data.table(char=c("1","2","3"), num=c(1,2,NA), list=list(1,NA,3)))
+  # measure supports cols arg, #5063
+  test(1035.0188, melt(DT_missing_l_2, measure.vars=measure(prefix, char, sep="_", cols=c("list_1","list_3"))), data.table(num_1=1,num_2=2,prefix="list",char=c("1","3"),value=list(1,3)))
+  test(1035.0189, melt(DT_missing_l_2, measure.vars=measure(value.name, char, sep="_", cols=c("list_1","list_3"))), data.table(num_1=1,num_2=2,char=c("1","3"),list=list(1,3)))
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:3164]: What happens when we specify both `patterns=` and `cols=`?

Comment (tdhock) [reply inst/tests/tests.Rraw:3164]: works as expected, same result, I just added a couple of tests.

Comment (tdhock) [reply inst/tests/tests.Rraw:3164]: so do you agree that the approach in this PR (supporting cols arg) is preferable to the other PR?

Comment (MichaelChirico) [reply inst/tests/tests.Rraw:3164]: I think I prefer this one. `patterns(cols=)` indeed already works as I'd expect:

```r
DT=data.table(a=1,b=2,c=3, aa=4)
DT[, .SD, .SDcols=patterns("^a", cols=c("a", "b"))]
#        a
#    <num>
# 1:     1
```

So deprecating `cols=` just removes a functionality, whereas this PR maintains that & adds new functionality.

Comment (tdhock) [reply inst/tests/tests.Rraw:3164]: agree

**Comment by tdhock** [comment]
ok tests are passing and I added a NEWS item, so I believe this one is ready to merge.

**Review by MichaelChirico — APPROVED** [review]
Excellent :)

**Pull request merged**